로그인 API 및 OAuth2 로그인 로직 추가

seonpilKim opened the pull request:
<!--
✅ Resolve: #이슈번호 형태로 입력해 주세요.
ex) Resolve: #123
-->
## 📌 Linked Issues
- Resolve: #11 

<!--
✅ 변경 사항을 자세히 알려주세요. (why -> what -> how)
-->
## 🔎 Change Details
### 1. OAuth2(kakao, naver) 로그인 로직 추가
- 첫 번째 로그인인 경우, 회원 가입 로직 포함(`CustomOAuth2UserService`)
- 로그인에 성공한 경우 ResponseBody에 JWT(AccessToken, RefreshToken)을 넣어서 전달(`OAuth2SuccessHandler`)
    - 개발 서버 배포 후에는 RefreshToken은 Cookie(HttpOnly)에 저장하도록 수정 필요

### 2. 인증/인가 실패 핸들러 추가 및 시큐리티 설정 클래스 추가
- `CustomAccessDeniedHandler`: 404 응답
- `CustomAuthenticationEntryPoint`: RequestHeader가 올바르지 않거나, JWT 관련 예외인 경우 401 응답
    > JWT 관련 예외가 여러 개 있는데, 몇 개는 정확하게 어떤 상황에서 발생하는지 모르겠어서 따로 ErrorCode는 추가하지 않았고, 각 예외 기본 메시지를 그대로 사용했습니다.
    > 이 부분은 천천히 좀 더 찾아보면서 업데이트하도록 하겠습니다.

### 3. Member 엔티티 필드명 변경
- email -> username: 추후에 username과 email을 분리할 가능성을 염두하여 수정하였습니다.
    - 회원 가입 시에만 `@Email`으로 유효성 검증을 진행하고, DB에는 username으로 저장하는 것이 로그인에 사용되는 아이디라는 것을 파악하기 쉬울 것 같다는 생각도 들었습니다.

### 4. 회원 가입 시 비밀번호 검증 정규식 수정
- 사용 가능한 특수 문자: `, ~, *, (, ) 추가
    - 안정성을 조금 더 높이기 위해 사용 가능한 특수 문자를 몇 개 더 추가하였습니다.

### 5. 이메일 로그인 기능 추가
- 로그인 요청 시 다음의 3가지 결과가 존재합니다.
    1. 해당 아이디가 존재하지 않는 경우 -> 200 응답
    2. 비밀번호가 일치하지 않는 경우 -> 200 응답
    3. 로그인에 성공한 경우 -> 200 응답
- 클라이언트가 API 스펙을 어긴 것이 아니기 때문에, 1번과 2번의 경우 200 응답으로 수정하였습니다.

### 6. 리스트 반환 시 불변 객체로 반환
- 회원의 권한을 조회할 때 리스트로 반환받는데, 이 리스트는 변경될 일이 없으므로 불변 객체로 감싸서 반환하도록 수정하였습니다.
    - 이렇게 사용하면 가비지 컬렉션의 성능 측면에서도 이점이 있다고 합니다.😲 ([Reference1](https://openjdk.java.net/jeps/269))
    > `Collections.unmodifiableList()` vs `List.copyOf()` vs `Arrays.asList()` vs `new ArrayList<>()`
    > `Collections.unmodifiableList()`: 수정 불가능한 리스트를 복사하지만, 원본과의 참조가 유지되어 원본이 수정되는 경우, 복사된 리스트의 값도 변경된다.
    > `List.copyOf()` : 원본과의 참조를 끊는 불변 리스트를 복사한다. 복사 대상이 불변 리스트인 경우 참조는 유지된다. (`JDK 9`)
    > `Arrays.asList()`: 원소의 수정은 가능하지만, 추가 및 제거는 불가능하다.
    > `new ArrayList<>()`: 원본과의 참조를 끊는 복사를 한다. 복사된 객체는 가변이다.

<!--
✅ 추가로 전달할 내용이 있다면 적어주세요.
-->
## 💬 Comment
- 시큐리티 관련 테스트는 천천히 작성해 보도록 하겠습니다!
- 회의 때 말씀드렸던 email 관련 질문입니다😇
    - 현재 JWT subject에 username(email)을 넣고 있는데, 카카오의 경우 이메일 수신 동의를 안하면 빈 값을 받게 됨 -> DB에서도 unique 속성을 만족 못함
        - 대안 1. username을 이메일 형식으로 사용x (필요 시 Member 테이블에 email 속성 추가)
        - 대안 2. JWT subject에 id(pk)를 넣어서 전달 + username(email) -> unique 속성 제거

<!--
✅ 참고한 문서가 있다면 공유해 주세요.
-->
## 📑 References
- (1) https://openjdk.java.net/jeps/269

## ✅ Check Lists
- [x] 추가한 기능에 대한 테스트는 모두 완료하셨나요?
- [x] 코드 정렬(Ctrl + Alt + L), 불필요한 코드나 오타는 없는지 확인하셨나요?
- [x] merge할 base branch가 올바른지 확인하셨나요?

seonpilKim committed:
Feat: Security + Jwt 인증 로직 추가

Resolve: #11

seonpilKim committed:
Chore: spring-boot-starter-oauth2-client 의존성 추가

Resolve: #11

seonpilKim committed:
Refactor: Member 엔티티 필드명 변경

email -> username

Resolve: #11

seonpilKim committed:
Feat: OAuth2 로그인/회원가입 로직 추가

Resolve: #11

seonpilKim committed:
Feat: SignupRequest 비밀번호 검증 정규식 수정

- 사용 가능한 특수 문자: `, ~, *, (, ) 추가

seonpilKim committed:
Feat: 이메일 로그인 기능 추가

- 도메인 및 HTTPS 적용 후, 재발급 토큰은 쿠키에 저장 필요

Resolve: #11

seonpilKim committed:
Feat: 이메일 로그인 로직 수정

다음의 경우 400 응답에서 200 응답으로 수정
- 회원 아이디가 존재하지 않는 경우
- 회원 비밀번호가 일치하지 않은 경우

signin() 반환 타입 변경
- 결과 상태를 추가로 전달
- 로그인 실패 시, jwt에는 빈 값을 추가

Resolve: #11

seonpilKim committed:
Refactor: 리스트 반환 시 불변 객체로 반환

회원 권한을 조회한 후 이를 수정할 일이 없기 때문에 불변 객체로 감싸서 반환.

Resolve: #11

seonpilKim committed:
Test: 로그인 로직 단위 테스트 코드 작성

상황: 로그인 성공, 아이디 존재x, 비밀번호 불일치

편의 메소드 추가
- getMember()
- getAuthorities()

Resolve: #11

seonpilKim committed:
Refactor: 테스트 유틸 클래스 추가

MemberUtils 전역 메소드 추가
- getMember()
- getAuthorities()

Resolve: #11

seonpilKim committed:
Test: MemberAuthorityService 단위 테스트 코드 작성

Resolve: #11

slolee commented on module-domain/src/main/java/com/zoopi/filter/JwtRequestFilter.java:
```diff
@@ -0,0 +1,76 @@
+
+@Component
+@RequiredArgsConstructor
+public class JwtRequestFilter extends OncePerRequestFilter {
```
Spring Security 자체에서 제공해주는 Security Filter (AuthenticationFilter) 가 별도로 있는걸로 알고있는데, 서블릿 필터에서 JWT 토큰 검증로직을 구현한 이유가 있을까요?

Spring Security 아키텍처대로 구현되지 않고 해당 필터에 너무 많은 내용이 들어오다보니 필터의 doFilterInternal 메소드의 책임이 너무 많아지고, 그에 따라서 try-catch 문과 if 문이 깊어지는 것 같습니다.

seonpilKim replied:
리뷰 감사드립니다 👍 

> Spring Security 자체에서 제공해주는 Security Filter (AuthenticationFilter) 가 별도로 있는걸로 알고있는데, 서블릿 필터에서 JWT 토큰 검증로직을 구현한 이유가 있을까요?

아 제가 알기로는 Security 라이브러리에 JWT 기반 인증 filter가 없는 걸로 알고 있어서 JwtRequestFilter를 만들어 UsernamePasswordAuthenticationFilter 뒤에 추가해 주었습니다!

> 해당 필터에 너무 많은 내용이 들어오다보니 필터의 doFilterInternal 메소드의 책임이 너무 많아지고, 그에 따라서 try-catch 문과 if 문이 깊어지는 것 같습니다.

SRP 원칙을 고려해서 리팩토링 해보겠습니다!

slolee replied:
JWT 기반 인증 필터가 직접적으로 제공되는것은 없지만 AbstractAuthenticationProcessingFilter 를 상속한 Filter 를 직접 만들어서 사용할 수 있습니다!

Spring Security 아키텍처에 대해서 제가 공부했던걸 간략하게 정리하면,

1. Security Filter 의 attemptAuthentication 에서 요청에 포함된 값(Authorization Header 등) 을 추출해 검증하고 Authentication 인증 전 객체로 만들고,
2. 이를 Provider 로 넘겨서 거기서 실제 인증 비즈니스 로직을 수행합니다. 
3. 인증이 성공적으로 됐다면 이를 Authentication 인증 완료 객체로 만들어 반환하고 Security Filter 의 successfulHandler, failuerHandler 에서 각각 후과정을 작업하도록되어있습니다.

이 때문에 각각의 책임이 분리되어 코드를 구성할 수 있다고 알고있습니다!
당연히 이 방법만이 정답인건 아니고 리팩토링하실때 고민대상으로 같이 고려해보면 어떨까 싶어서 공유드립니다!

dongkyunkimdev replied:
찬준님이 말씀하신 것처럼 Spring Security 아키텍처에 기반하여 Provider 등으로 리팩토링 되어질 여지가 충분히 있다고 생각하구요, 
```
errors.add(new ErrorResponse.FieldError("Authorization", "",
				AUTHORIZATION_HEADER_MISSING.getMessage()));
```
위 코드에서 하드코딩된 문자열들도 추후에 AUTHORIZATION_HEADER처럼 의미있는 이름을 붙여 사용하면 더 가독성 좋은 코드가 될 것 같고, 이런 것들 차후에 이슈 별도로 따서 리팩토링 하면 좋을 것 같습니다!
전반적으로 인증 구현을 잘 하신 것 같아요 역시 선필님!! 고생하셨습니다👏

seonpilKim replied:
두 분 모두 정말 감사합니다! 다음 리팩토링 때 같이 적용해 보겠습니다 👍👍

slolee commented:
1. 개발 서버 배포 후에는 RefreshToken은 Cookie(HttpOnly)에 저장하도록 수정 필요
 -> 저 같은 경우에는 노출시 위험한 정보인 RefreshToken 같은 경우에는 왠만하면 쿠키에 저장하지 않는 것이 좋다고 알고있습니다. 따라서 단순히 Response Body 에 담아서 반환하고 프론트쪽에서 메모리에 저장할 수 있도록 가이드하는편입니다!

seonpilKim commented:
@slolee 
앗 제가 알고 있는 거랑 살짝 다른 것 같은데, 혹시 제가 생각한 내용을 정리해 드릴테니 틀린 내용이 있으면 지적 부탁드려도 될까요?!

우선 JWT(AccessToken + RefreshToken) 기반 인증 방식은 토큰의 탈취 위험이 있어서 이를 보완하는 게 중요한 것으로 알고 있는데요, AccessToken는 직접 인증 처리를 위한 토큰이라 가장 안전한 곳(메모리)에 저장하는 것으로 알고 있습니다.

다만, 네트워크 상에서의 탈취나 시스템 해킹을 통해 탈취될 위험도 어쩔 수 없이 존재하기 때문에, AccessToken의 만료 시간을 짧게 만드는 방식으로 대응하는 것으로 알고 있습니다. 
(저는 10~30분? 정도로 하면 어떨까 싶은데 다른 분들의 의견도 궁금합니다!)

또한 RefreshToken은 AccessToken이 만료되었을 때 사용자가 다시 재로그인하는 과정을 생략하기 위해, RefreshToken을 서버에 전달하여 새로운 AccessToken과 RefreshToken을 발급받는 방식으로 사용자의 인증 상태가 유지될 수 있는 목적으로 사용하는 것으로 알고 있습니다.

따라서 RefreshToken은 클라이언트가 계속 보관하고 있어야 나중에 재발급 요청을 할 수 있기 때문에, 가능한 안전하게 유지될 수 있는 공간에 저장해야 하는 것이 중요한데, 이 때 브라우저의 Local Storage나 Session Storage에 저장하면 XSS 공격에 의해 탈취당할 가능성이 매우 높고, 메모리에 저장하면 SPA 방식에서 어느 정도 유지가 될 수는 있지만 클라이언트가 새로 고침을 한다던지, 브라우저를 껏다가 다시 키는 상황에는 다시 사라지기 때문에 재로그인이 필요한 불편함이 생깁니다.

그래서 나온 대안이 Cookie에 저장하는 것으로 알고 있고, 쿠키 또한 탈취 위험이 있으나 httpOnly 옵션으로 http 통신을 통해서만 쿠키를 꺼낼 수 있게 보호하고, secure 옵션으로 네트워크 상에서 탈취당해도 기밀성이 보장되게 하며, domain 옵션과 sameSite 옵션으로 CSRF 공격도 어느 정도 대비하는 방식으로 사용하는 것으로 알고 있습니다. 

[Reference](https://codeburst.io/localstorage-vs-cookies-all-you-need-to-know-about-storing-jwt-tokens-securely-in-the-front-end-70dc0a9b3ad3)

slolee commented:
작성해주신 내용 꼼꼼히 읽어보았습니다! 좋은 참고자료도 감사드립니다 :D

궁금한점은 아래 두가지입니다.
1. RefreshToken 또한 노출되었을때 공격자가 이를 이용해 AccessToken, RefreshToken 을 재발급해 인증정보를 도용할 수 있지 않을까요?
   (CORS 를 통해서 어느정도 보완이 가능하긴 하지만 애초에 노출되는 것을 방지하는 것이 더 높은 보안수준일 것이라 생각합니다)
2. RefreshToken 을 메모리에 저장할경우 새로고침시 초기화되는 문제는 redux-persist 와 같은 라이브러리를 통해 해결할 수 있을것 같은데(브라우저 닫는경우 해당x) 어떻게 생각하시나요?

seonpilKim commented:
@slolee 

> RefreshToken 또한 노출되었을때 공격자가 이를 이용해 AccessToken, RefreshToken 을 재발급해 인증정보를 도용할 수 있지 않을까요?
(CORS 를 통해서 어느정도 보완이 가능하긴 하지만 애초에 노출되는 것을 방지하는 것이 더 높은 보안수준일 것이라 생각합니다)

음 먼저 RefreshToken이 노출될 수 있는 상황들을 구체화할 수 있다면 좋은 해결책이 나올 수 있지 않을까 싶습니다!
그래서 일단 저희가 적용할 수 있는 방안들을 정리해보면
1. Cookie에 httpOnly 옵션을 적용하면, http 통신을 통해서만 쿠키 내용을 확인할 수 있다. (XSS도 방어 가능)
    - 사용자의 브라우저에서도 쿠키 확인이 가능하지만, 이는 시스템 해킹이나 사용자 PC를 몰래 염탐하지 않는 이상 쿠키 노출 위험은 없을 거라 생각됩니다!
2. Cookie에 secure 옵션을 적용하면, 네트워크 상에서 패킷이 탈취되더라도 secret key가 없는 이상 쿠키 내용을 복호화 할 수 없다.
3. Cookie에 domain 및 sameSite 옵션을 사용하면, 지정한 도메인이 아닌 이상 쿠키 전달이 불가능하게 할 수 있다.
    - 쿠키가 탈취되더라도, 저희가 지정한 도메인 주소가 아닌 이상 해당 쿠키로 HTTP 요청을 보내도 쿠키 전달이 불가능
    - 따라서 재발급 토큰을 탈취한다 하더라도, CSRF 공격을 통해 우회하는 방식으로 HTTP 요청을 보내는 방법 외에는 토큰 재발급을 받을 수 없음 -> 위의 옵션으로 CSRF 공격 대비 가능
    - CORS만으로는 CSRF 공격을 막을 수 없지만, 위의 옵션을 이용하면 방어가 가능합니다.

정리해 보자면, RefreshToken이 노출될 수 있는 상황은 시스템 해킹 or 사용자 PC 염탐 정도 인 것 같다고 생각합니다!
혹시 또 다른 상황이 있을까요?!

> RefreshToken 을 메모리에 저장할경우 새로고침시 초기화되는 문제는 redux-persist 와 같은 라이브러리를 통해 해결할 수 있을것 같은데(브라우저 닫는경우 해당x) 어떻게 생각하시나요?

제가 알기로 redux-persist는 브라우저의 local-storage 혹은 session-storage에 저장하는 것으로 알고 있는데, 여기에 저장하면 XSS 공격으로 탈취가 가능한 것으로 알고 있습니다!

slolee commented:
오.. 구체적인 답변 감사합니다..!
일단 말씀해주신 내용 이외에 저도 떠오르는부분은 없는거 같습니다.
토큰 저장관련해서는 공유주신 내용으로 진행해도 될것 같습니다! 좋은 의견 감사드립니다~! :D

seonpilKim commented:
AWS 관련해서 댓글 남긴다는 걸 깜박했네요😅
한 인스턴스에 프론트 + 백엔드 서버를 같이 배포하면, 호스트가 동일하다보니 따로 도메인을 구매하지 않고도 쿠키 전달이 가능합니다! (httpOnly + domain 옵션 사용)
나중에 도메인도 구매하고 HTTPS도 적용하면 추가로 secure 옵션과 sameSite = LAX 옵션도 추가하면 될 것 같습니다 :)

dongkyunkimdev commented on module-domain/src/main/java/com/zoopi/domain/member/service/MemberService.java:
```diff
@@ -17,18 +30,53 @@ public class MemberService {
+			.member(member)
+			.type(AuthorityType.ROLE_USER)
+			.build();
+		memberAuthorityRepository.save(memberAuthority);
```
위 멤버 권한 저장하는 부분 private 메서드로 extract하는 것도 괜찮을 것 같습니다!

장점으로는,
> 메소드가 잘 쪼개져 있을 경우 재사용성이 높아짐
> 고수준의 메서드를 볼 때 일련의 주석을 읽는 것 같은 느낌이 들도록 할 수 있음

예시,
```
@Transactional
public Member createMember(String email, String phone, String name, String password, JoinType joinType) {
	final Member member = Member.builder()
		.username(email)
		.password(passwordEncoder.encode(password))
		.name(name)
		.phone(phone)
		.joinType(joinType)
		.build();
	memberRepository.save(member);

	saveMemberAuthority(member);

	return member;
}

private void saveMemberAuthority(Member member) {
	final MemberAuthority memberAuthority = MemberAuthority.builder()
		.member(member)
		.type(AuthorityType.ROLE_USER)
		.build();
	memberAuthorityRepository.save(memberAuthority);
}
```

seonpilKim replied:
좋은 생각인 것 같아요! 바로 적용하겠습니다🐣

dongkyunkimdev commented on module-domain/src/test/java/com/zoopi/domain/member/service/MemberAuthorityServiceTest.java:
```diff
@@ -0,0 +1,84 @@
+		final List<MemberAuthority> authorities = getAuthorities(member, AuthorityType.ROLE_USER);
+
+		doReturn(Optional.of(member)).when(memberRepository).findByUsername(member.getUsername());
+		doReturn(authorities).when(memberAuthorityRepository).findAllByMember(member);
```
테스트코드를 작성할 때 BDD의 권장 행동 패턴인 Given/When/Then을 사용하는 경우,
given 단계에서 when 메서드가 들어가는 등 가독성에 혼돈을 주는 Mockito를 해결하기 위해 나온 BDDMockito를 사용하는 것도 고려해볼 만한 것 같습니다!
참고 : [Mockito와 BDDMockito는 뭐가 다를까?](https://velog.io/@lxxjn0/Mockito%EC%99%80-BDDMockito%EB%8A%94-%EB%AD%90%EA%B0%80-%EB%8B%A4%EB%A5%BC%EA%B9%8C)

seonpilKim replied:
와..!! `doReturn().when()`쓰면서 항상 가독성이 눈에 걸렸었는데, BDDMockito의 `given().willReturn()`를 사용하면 훨씬 가독성있게 보일 것 같네요!
좋은 정보 알려주셔서 감사합니다👍 이건 따로 이슈 남겨서 리팩토링 하겠습니다!

dongkyunkimdev commented on module-domain/src/test/java/com/zoopi/domain/member/util/MemberUtils.java:
```diff
@@ -0,0 +1,33 @@
+import com.zoopi.domain.member.entity.Member;
+import com.zoopi.domain.member.entity.MemberAuthority;
+
+public class MemberUtils {
```
👍

seonpilKim replied:
☺

seonpilKim committed:
Refactor: 메소드 추출

saveMemberAuthority()
- 회원 가입 과정에서 ROLE_USER 권한을 DB에 추가하는 메소드

generateJwt()
- 회원 정보과 권한을 이용해 AccessToken과 RefreshToken을 생성하여 반환하는 메소드

Resolve: #11
